bc_obps/compliance/service/penalty_calculation_service.py

```diff
@@ -327,4 +450,174 @@ def create_penalty(cls, obligation: ComplianceObligation) -> CompliancePenalty:
             final_accrual_date=final_transaction_date,
         )
 
-        return CompliancePenalty.objects.get(compliance_obligation=obligation)
+        return CompliancePenalty.objects.get(
+            compliance_obligation=obligation,
+            penalty_type=CompliancePenalty.PenaltyType.AUTOMATIC_OVERDUE,
+        )
+
+    @classmethod
+    def _get_interest_rate_resolver(
+        cls, accrual_start_date: date, final_accrual_date: date
+    ) -> Callable[[date], Decimal]:
+        """
+        Return an interest rate resolver for the accrual period.
+        """
+        interest_rates = ElicensingInterestRate.objects.filter(
+            start_date__lte=final_accrual_date,
+            end_date__gte=accrual_start_date,
+        ).order_by('-start_date')
+
+        if not interest_rates.exists():
+            interest_rates = ElicensingInterestRate.objects.filter(is_current_rate=True)
+
+        if not interest_rates.exists():
+            raise ValueError("No interest rates configured in the system")
+
+        def get_rate_for_date(date_to_check: date) -> Decimal:
+            m = date_to_check.month
+            y = date_to_check.year
+            match m:
+                case 1 | 2 | 3:
+                    reference_date = date(y - 1, 12, 15)
+                case 4 | 5 | 6:
+                    reference_date = date(y, 3, 15)
+                case 7 | 8 | 9:
+                    reference_date = date(y, 6, 15)
+                case 10 | 11 | 12:
+                    reference_date = date(y, 9, 15)
+
+            for rate in interest_rates:
+                if rate.start_date <= reference_date <= rate.end_date:
+                    return rate.interest_rate
+            return ElicensingInterestRate.objects.filter(is_current_rate=True).first().interest_rate  # type: ignore[union-attr]
+
+        return get_rate_for_date
```

Comment:
I think we can simplify this function:
I don't think we need the resolver function at all & can just define the get_rate_for_date() function that checks the date of compounding and returns a .get() from the ElicensingInterestRate table for the matching reference_date
```suggestion
      def get_rate_for_date(date_to_check: date) -> Decimal:
          m = date_to_check.month
          y = date_to_check.year
          match m:
              case 1 | 2 | 3:
                  reference_date = date(y - 1, 12, 15)
              case 4 | 5 | 6:
                  reference_date = date(y, 3, 15)
              case 7 | 8 | 9:
                  reference_date = date(y, 6, 15)
              case 10 | 11 | 12:
                  reference_date = date(y, 9, 15)
          return ElicensingInterestRate.objects.get(start_date__lte=reference_date,end_date__gte=reference_date).interest_rate
```
This removes a bunch of unnecessary queries that check for the existence of interest rates before the date matching & still fails appropriately if an interest rate does not exist for the .get() query as there should always be 1 and only 1 interest rate record match for a reference_date